seqr/views/utils/terra_api_utils.py

```diff
@@ -0,0 +1,227 @@
+    r = session.get("register")
+    if r.status_code != 200:
+        raise TerraAPIException(
+            'Error: called Terra API "register" got status: {} with a reason: {}'.format(r.status_code, r.reason))
+    return json.loads(r.text)
+
+
+def list_anvil_workspaces(user, fields=None):
+    """
+    Get all the workspaces accessible by the logged-in user.
+
+    Args:
+    user (User model): who's credentials will be used to access AnVIL
+    fields (str): a comma-delimited list of values that limits the
+        response payload to include only those keys and exclude other
+        keys (e.g., to include {"workspace": {"attributes": {...}}},
+        specify "workspace.attributes").
+    """
+    session = anvil_session_store.get_session(user)
+    if fields is None:
+        r = session.get("api/workspaces")
+    else:
+        r = session.get("api/workspaces", params = {"fields": fields})
+    if r.status_code != 200:
+        raise TerraAPIException(
+            'Error: called Terra API "api/workspaces" got status: {} with a reason: {}'.format(r.status_code, r.reason))
+    return json.loads(r.text)
+
+
+def get_anvil_workspace_acl(workspace_namespace, workspace_name):
+    """
+    Request FireCloud access control list for workspace.
+
+    Args:
+        workspace_namespace (str): namespace (name of billing project) of the workspace
+        workspace_name (str): the name of the workspace
+    Returns:
+        {
+            "user1Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            },
+            "user2Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            },
+            "user3Email": {
+              "accessLevel": "string",
+              "pending": true,
+              "canShare": true,
+              "canCompute": true
+            }
+          }
+          :param workspace_name:
+          :param workspace_namespace:
+    """
+    if workspace_namespace and workspace_name:
+        uri = "api/workspaces/{0}/{1}/acl".format(workspace_namespace, workspace_name)
+        r = anvil_session_store.service_account_session.get(uri)
+        if r.status_code in DEFAULT_REFRESH_STATUS_CODES:  # has failed in refreshing the access code
+            anvil_session_store.service_account_session = AnvilSession(
+                service_account_info = GOOGLE_SERVICE_ACCOUNT_INFO, scopes = SOCIAL_AUTH_GOOGLE_OAUTH2_SCOPE)
+            r = anvil_session_store.service_account_session.get(uri)  # retry with the new access code
+        if r.status_code != 200:
+            raise TerraAPIException(
+                'Error: called Terra API "{}" got status: {} with a reason: {}'.format(uri, r.status_code, r.reason))
+        return json.loads(r.text)['acl']
+    else:
+        return {}
```

Comment:
It is okay to check the workspace name before calling this function. It is not an error because it is normal for some projects to have no workspaces.

Reply:
but if a project has no workspace you shouldn't call this function at all

Reply:
I changed it to check the workspace existing before calling this function. If somebody uses this function without checking, the function will make a call to AnVIL for workspace `None/None`. And it will get a response with a 404 status. Then, the function will raise an exception since the status is not 200. Is that okay?

Reply:
So you shouldn't gracefully handle the case where theres no workspace in this function the way you do here with this if/else. Either you should be checking before it is called and then inside this function don't do any validation or error handling, or you should check and explicitly throw an error in this function if theres no workspace. If you go with the option of assuming you only call this function when theres a workspace, its okay if its miss-used the way you described, but you shouldn't account for that in the code

Reply:
thanks for cleaning this up, its much clearer to me what is happening now and what we will need to fix once we are whitelisted. It looks like this is the only place in the code that we will need to make a change once we are whitelisted - is that correct?

Reply:
Yes, I think so.